python/ray/data/dataset.py

```diff
@@ -2858,6 +2859,7 @@ def iter_batches(
                 to select ``numpy.ndarray`` for tensor datasets and
                 ``Dict[str, numpy.ndarray]`` for tabular datasets. Default is "default".
             drop_last: Whether to drop the last batch if it's incomplete.
+            collate_fn: A function to apply to each data batch before returning it.
```

Comment:
Hmm I don't know if it makes sense to have this as part of the `ds.iter_batches()` API, since the primary motivation behind this feature is to (1) make the Torch `DataLoader` port easier, and (2) provide a hook to override how we convert NumPy ndarrays to Torch Tensors. `collate_fn` doesn't provide much critical value for the core `ds.iter_batches()` API: it's functionally exactly equivalent to
```python
for batch in ds.iter_batches():
    batch = collate_fn(batch)
```

Is the primary motivation for this to be able to include it in the local pipelining operation chain, to improve collation performance? If that's the case, I'm wondering if `ds.iter_torch_batches()` shouldn't reuse `ds.iter_batches()`, and we should instead have `ds.iter_torch_batches()` can use the internal `batch_block_refs` API directly, which can expose `collate_fn`. Most of the batching logic is captured in `batch_block_refs` anyway.

Reply:
Yeah I was also thinking the same...will make the change.

Easiest thing is just to have a private arg in `iter_batches`. IMO that would be preferable for future maintenance than both APIs directly using `batch_block_refs`

Reply:
Updated to make `_collate_fn` private for `iter_batches`

Reply:
Hmm I'd rather not muddy the user-facing `.iter_batches()` API even with a private `_collate_fn` arg, if possible, since it's already a pretty wide interface. And since `batch_block_refs` contains most of the batching logic, I don't think that there's much maintenance or drift risk here.

What are your primary concerns for maintenance here?

Reply:
it's mostly about the timing logic, but it's not a big deal. Updated to using `batch_block_refs` directly.

Reply:
If stuff like execution triggering and stats collection becomes involved enough around `batch_block_refs`, we can always introduce a private `Dataset._iter_batches()` method that can support all of the other batch-based iterator APIs without making the user-facing `Dataset.iter_batches()` API any wider.

Reply:
DatasetPipeline makes an implicit assumption that `Dataset.iter_torch_batches` calls into `self.iter_batches`

Reply:
I think private arg is the way to go, because the abstractions here are very messy. We can remove this later once we deprecate `DatasetPipeline`.